### DISCUSSION TITLE
HW 01, Problems Part 2, Question 1: Combining Strings and Variables

### DISCUSSION THREAD

**Discussion by feliciones** [cooperrc/computational-mechanics#21] [Ideas]
Hello,

I was recently going through the first HW and I came across a useful line of code using an f-string that allows you to combine strings and variables. I wanted to paste my code below in case you were trying to get a similar output. See it below.

![image](https://user-images.githubusercontent.com/106564968/171304023-135724e9-4a26-4157-a58f-f25bd4fe80ef.png)

I hope this helps,
Simon

**Comment by UniquelyAbbi** [comment]
@feliciones I am confused on that part of the HW. I did the def sincos(x) inout but I’m not sure about the rest I’m looking at the notebook but I still don’t understand it. I get what yours is saying though. I’ll go back look more at the book. Thanks for the help

**Reply by feliciones** [reply]
I am still attempting to work through the second part of that question and graphing it. I was sharing my code as an example on the 'print' line to show how to do it. I don't believe I wrote my code entirely efficient to code the result, so I will return to that tonight and can let you know. But if anyone has any suggestions for the plotting that would also be greatly appreciated.

**Reply by Aidan-Myatt** [reply]
For how I solved this problem, I had the function sincos(x) return np.sin(x) and np.cos(x). From there, I assigned two variables to the function by writing y1, y2 = sincos(x). You then use np.linspace to create an array for the x axis ranging from 0 to 2pi. After that, plot it like normal.

**Reply by ecieszynski** [reply]
Aidan, i just tried something similar to that but the plot shows nothing. This is my code. Any guidance would be appreciated

![image](https://user-images.githubusercontent.com/106542680/171485894-f4e0a5d0-1a1e-4d6f-874d-40b152593773.png)

**Reply by Aidan-Myatt** [reply]
Your function is defined correctly. You just have a small error when assigning it. Try plugging this in instead after the function:
x_range = np.linspace(0, 2*np.pi)
y1, y2 = sincos(x_range)

After you plug that in, you should be all set. I think you just misunderstood the problem as it was worded weirdly. It should be plotted from 0 to 2pi instead of at x = 0.2 specifically. Let me know if that works.

**Reply by ecieszynski** [reply]
Still seeing nothing

![image](https://user-images.githubusercontent.com/106542680/171513656-a1f772e4-a65c-4798-a0ba-d4e0a2e00e56.png)

**Reply by feliciones** [reply]
I made a similar error, what I forgot to do was to make sure to define the range x could be in before writing the function (see below).

![image](https://user-images.githubusercontent.com/106564968/171518461-7e4b0568-c199-4ae8-b1c7-f2d9d934f9e7.png)

**Reply by Aidan-Myatt** [reply]
That shouldn't be the problem from my understanding. I honestly have no idea why your code is not working. This is what I put for reference. 

![image](https://user-images.githubusercontent.com/106316906/171521422-2af12691-3a3e-41b7-90d0-f54d7c98f42e.png)

**Reply by FishMacaron** [reply]
@ecieszynski I think I see your problem! For your x_range value, try x_range=(0,2\*np.pi) instead of x_range=(0,2\*np.pi,1).
I had a similar issue when I put in three arguments and that did the trick for me when nothing was showing up.

**Reply by ecieszynski** [reply]
this solved it! Could explain the reasoning why that worked and the way I had it didn't? Just want to better understand where I went wrong.

**Reply by cooperrc** [reply]
Nice job!!

Explaining the [`linspace`](https://numpy.org/doc/stable/reference/generated/numpy.linspace.html) function

```python
linear_range = np.linspace(start, end, steps)
```

- `start` is the initial value in the array, `linear_range`
- `end` is the final value in the array, `linear_range`
- `steps` is the length of `linear_range`

so if you call, `np.linspace(0, 2*np.pi, 1)` you get an array with length of 1:
`array([0])`

the default is length of 50 so you get

`array([0.        , 0.12822827, 0.25645654, ..., 6.28318531])`

**Reply by UniquelyAbbi** [reply]
I have been trying to get this part on my own but it isn't working out.  It keeps saying that my np.linespace(0, 2*np.pi) is wrong i believe

AttributeError: module 'numpy' has no attribute 'linespace'

**Reply by feliciones** [reply]
Make sure you imported numpy at the beginning as well as making sure it’s np.linspace and not linespace with the “e”

**Reply by cooperrc** [reply]
> I have been trying to get this part on my own but it isn't working out. It keeps saying that my np.linespace(0, 2*np.pi) is wrong i believe
> 
> AttributeError: module 'numpy' has no attribute 'linespace'

Its a typo in how you're calling the function its *lin*space, short for *lin*ear space:

~~`linespace`~~
->
`linspace`

**Reply by UniquelyAbbi** [reply]
omg thank you. I have been adding the e to make it line space.
